core/modules/replica/Controller.h

```diff
@@ -0,0 +1,687 @@
+#include "replica/ServiceProvider.h"
+
+// Forward declarations
+
+// This header declarations
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+// Forward declarations
+
+class ControllerImpl;
+class Messenger;
+
+/**
+ * The base class for implementing requests registry as a polymorphic
+ * collection to store active requests. Pure virtual methods of
+ * the class will be overriden by request-type-specific implementations
+ * (see struct RequestWrappeImplr<REQUEST_TYPE> in the .cc file) capturing
+ * type-dependant pointer and a callback function.
+ */
+struct ControllerRequestWrapper {
+
+    /// The pointer type for instances of the class
+    typedef std::shared_ptr<ControllerRequestWrapper> pointer;
+
+    /// Destructor
+    virtual ~ControllerRequestWrapper() = default;
+
+    /// This method will be called upon a completion of a request
+    /// to notify a subscriber on the event.
+    virtual void notify ()=0;
+
+    /// Return a pointer to the stored request object
+    virtual std::shared_ptr<Request> request () const=0;
+};
+
+/**
+ * The data structure encapsulating various attributes which identity
+ * each instance of the Controller class. This information is meant to
+ * be used in the multi-Controller setups to coordinate operations
+ * between multiple instances and to avoid/resolve conflicts.
+ */
+struct ControllerIdentity {
+
+    /// A unique identifier of the Controller
+    std::string id;
+
+    /// The name of a host where it runs
+    std::string host;
+
+    /// An identifier of a process
+    pid_t pid;
+};
+
+/// The overloaded streaming operator for Controller's identity
+std::ostream& operator <<(std::ostream& os, ControllerIdentity const& identity);
+
+/**
+  * Class Controller is used for pushing replication (etc.) requests
+  * to the worker replication services. Only one instance of this class is
+  * allowed per a thread.
+  *
+  * NOTES:
+  *
+  * - all methods launching, stopping or checking status of requests
+  *   require that the server was runinng. Otherwise it will throw
+  *   std::runtime_error. The current implementation of the server
+  *   doesn't support (yet?) amn operation queuing mechanism.
+  *
+  * - methods which take worker names as parameters will throw exception
+  *   std::invalid_argument if the specified worker names were not found
+  *   in the configuration.
+  */
+class Controller
+    :   public std::enable_shared_from_this<Controller> {
+
+public:
+
+    /// Friend class behind this implementation must have access to
+    /// the private methods
+    friend class ControllerImpl;
+
+    /// The pointer type for instances of the class
+    typedef std::shared_ptr<Controller> pointer;
+
+    /**
+     * Static factory method is needed to prevent issue with the lifespan
+     * and memory management of instances created otherwise (as values or via
+     * low-level pointers).
+     *
+     * @param serviceProvider - for configuration, other services
```

Comment:
is there more to this description?

Reply:
This is just a safest way to created objects for classes inheriting from:
std::enable_shared_from_this

Reply:
I don't understand, "other services", related to what? a more complete sentence would help, IMHO

Reply:
In the present design, the ServiceProvider class provides a number of services for the Replication system, which includes: the Configuration service, Database (persistency) service, Chunk Locker, etc. This list may grow i the future. Hence, I decided just to give here a hint of what the ServiceProvide is for w/o trying to list all possible services it has. It would be kind of silly to find all documentation entries where a reference/pointer to the ServiceProvider is passed and mention the names/meanings of all services :-)